### ISSUE TITLE
Where is state kept in the container?

### ISSUE THREAD

**Issue by arunderwood** [SumoLogic/sumologic-collector-docker#56]
Currently, every time I `docker rm sumologic` and `docker run sumologic/collector:latest-no-source`, the collector resends all my `LocalFile` sources up to Sumologic, creating a bunch of duplicate logs.

Is there a place in the container that tracks the state of what messages have been synced up that I could mount in a volume to persist sync state between container instances?

**Comment by maimaisie** [comment]
Hi. The collector has internal state to keep track of what has been collected and the state is persistent for container restarts but not redeploys, so we recommend not redeploying the collector container unless you need a newer collector version (we don’t release new versions very frequently though.) 

Another option is to install and run the collector as a service on your docker host. Unless you uninstall the service, the state is persistent through shutdown, restart and upgrade.

**Comment by arunderwood** [comment]
My goal is to make collector state persist through redeploys.  Is there a specific directory in the container that holds the state so I can put it on a docker volume?

**Comment by maimaisie** [comment]
The collector directory is `/opt/SumoCollector/` and contains collector/source configuration and states. However it is not officially verified that collector can persist through redeploys by putting the directory on a volume.

**Comment by nhoughto** [comment]
@arunderwood Did you attempt / succeed with this? We are seeing the same behaviour, its very annoying and probably should be mentioned in the README

**Comment by arunderwood** [comment]
No, sorry, I never dug into this.  I just stopped using the docker collector.

**Comment by ragnarkurm** [comment]
[Where does fingerprint data information gets stored?](https://support.sumologic.com/hc/en-us/articles/360035953493-Where-does-fingerprint-data-information-gets-stored-) This is where the state is kept.

In my case the folder `/opt/SumoCollector/config/blades` has following content:
```
00000000095079AA.fallback.json
00000000095079AA.json
00000000095079AB.fallback.json
00000000095079AB.json
00000000095079AC.fallback.json
00000000095079AC.json
```

Not sure if the naming is unique per deployment/execution or not.

**Comment by ragnarkurm** [comment]
Tried to share the `/opt/SumoCollector/config/blades` over aws efs mount.
Didn't work. Reason unknown.

**Comment by ragnarkurm** [comment]
SUCCESS, got it working. (with couple of downsides).

**How it works?**

1. I moved whole `/opt/SumoCollector/config` to persistent storage, and link it (`ln -s`) from original place.
2. Making sure no parallel execution happens during deploy, for example. `flock` works over NFS.

**Downsides**

The hostname. I seem not to control that. In my case collector name is `xxx-a149d5664cc7`. The ID (`a149d5664cc7`) seems to be constant across deployments. I cannot understand where it comes from or how to control that. I tried changing blade json files and /etc/hostname before controller starts, but no avail.

The sources update. If you updates your `sumo-sources.json` file, the changes may not be taken into account. At least need to change from web UI: Data Manage > Collection > the collector > Edit > Advanced > Local Configuration File.
